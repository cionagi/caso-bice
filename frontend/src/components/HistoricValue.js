/* eslint-disable react-hooks/exhaustive-deps */
// Dependency
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

// Components
import HistoricTable from './HistoricTable';
import Loading from './Loading';
import Pagination from './Pagination';
import LineChartContainer from './LineChartContainer';

const HistoricValue = ({ historicData, show }) => {
  const [state, setState] = useState({
    loading: true,
    data: [],
    dataShow: [],
    perPage: 20,
    pagination: {
      current: 0,
      total: 0,
    },
  });

  useEffect(() => {
    const {
      perPage,
      pagination: { current },
    } = state;

    setState((prevState) => ({
      ...prevState,
      loading: historicData.length,
      data: [...historicData],
      dataShow: historicData.slice(current * perPage, (current + 1) * perPage),
      pagination: {
        current: 0,
        total: Math.round(historicData.length / state.perPage),
      },
    }));
  }, [historicData]);

  useEffect(() => {
    const {
      data,
      perPage,
      pagination: { current },
    } = state;

    setState((prevState) => ({
      ...prevState,
      dataShow: data.slice(current * perPage, (current + 1) * perPage),
    }));
  }, [state.pagination.current]);

  const handlePagination = (page) => {
    setState({
      ...state,
      pagination: {
        ...state.pagination,
        current: page,
      },
    });
  };

  return (
    <div className={` mt-3 collapse ${show ? 'show' : undefined}`}>
      <div className="card card-body">
        {state.loading ? <HistoricTable historicData={state.dataShow} /> : <Loading />}

        <div className="row justify-content-center">
          <div className="col-12 col-md-8 mt-3 text-center d-flex justify-content-center">
            <Pagination
              current={state.pagination.current}
              totalPage={state.pagination.total}
              handlePagination={handlePagination}
            />
          </div>
        </div>
      </div>
      <div className="row">
        <div className="col-12 col-md-12">
          {!!historicData.length && <LineChartContainer historicData={historicData} />}
        </div>
      </div>
    </div>
  );
};

HistoricValue.defaultProps = {
  historicData: [],
};

HistoricValue.propTypes = {
  historicData: PropTypes.array,
  show: PropTypes.bool,
};

export default HistoricValue;
