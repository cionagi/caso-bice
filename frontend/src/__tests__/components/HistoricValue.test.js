import React from 'react';
import { render } from '@testing-library/react';

import HistoricValue from '../../components/HistoricValue';

test('Link renders correctly', () => {
  const { asFragment } = render(<HistoricValue />);
  expect(asFragment()).toMatchSnapshot();
});
