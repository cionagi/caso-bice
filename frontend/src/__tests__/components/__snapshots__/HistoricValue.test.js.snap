// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly 1`] = `
<DocumentFragment>
  <div
    class=" mt-3 collapse undefined"
  >
    <div
      class="card card-body"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <div
          class="spinner-border"
          role="status"
          style="width: 5rem; height: 5rem;"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
      <div
        class="row justify-content-center"
      >
        <div
          class="col-12 col-md-8 mt-3 text-center d-flex justify-content-center"
        >
          <nav>
            <ul
              class="pagination pagination-sm"
            />
          </nav>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-12 col-md-12"
      />
    </div>
  </div>
</DocumentFragment>
`;
